internal/nix/nixprofile/item.go

```diff
@@ -0,0 +1,67 @@
+package nixprofile
+
+import (
+	"fmt"
+	"strings"
+
+	"go.jetpack.io/devbox/internal/devpkg"
+	"go.jetpack.io/devbox/internal/lock"
+	"go.jetpack.io/devbox/internal/redact"
+)
+
+// NixProfileListItem is a go-struct of a line of printed output from `nix profile list`
+// docs: https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-profile-list.html
+type NixProfileListItem struct {
+	// An integer that can be used to unambiguously identify the package in
+	// invocations of nix profile remove and nix profile upgrade.
+	index int
+
+	// The original ("unlocked") flake reference and output attribute path used at installation time.
+	// NOTE that this can be "#" if the package was added to the nix profile via store path.
+	unlockedReference string
+
+	// The locked flake reference to which the unlocked flake reference was resolved.
+	// NOTE that this can be "#" if the package was added to the nix profile via store path.
+	lockedReference string
+
+	// The store path(s) of the package.
+	nixStorePath string // TODO: change to slice
+}
+
+// AttributePath parses the package attribute from the NixProfileListItem.lockedReference
+//
+// For example:
+// if NixProfileListItem.lockedReference = github:NixOS/nixpkgs/52e3e80afff4b16ccb7c52e9f0f5220552f03d04#legacyPackages.x86_64-darwin.go_1_19
+// then AttributePath = legacyPackages.x86_64-darwin.go_1_19
+func (item *NixProfileListItem) AttributePath() (string, error) {
+	// lockedReference example:
+	// github:NixOS/nixpkgs/52e3e80afff4b16ccb7c52e9f0f5220552f03d04#legacyPackages.x86_64-darwin.go_1_19
+
+	// AttributePath example:
+	// legacyPackages.x86_64.go_1_19
+	_ /*nixpkgs*/, attrPath, found := strings.Cut(item.lockedReference, "#")
+	if !found {
+		return "", redact.Errorf(
+			"expected to find # in lockedReference: %s from NixProfileListItem: %s",
+			redact.Safe(item.lockedReference),
+			item,
+		)
+	}
+	return attrPath, nil
+}
+
+// ToPackage constructs a nix.Package using the unlocked reference.
+// TODO: this doesn't handle well the case where item.unlockedReference == "#"
+func (item *NixProfileListItem) ToPackage(locker lock.Locker) *devpkg.Package {
+	return devpkg.PackageFromString(item.unlockedReference, locker)
```

Comment:
hmm this could be problematic for Items created via store-paths! 😬

Reply:
Seems to be used in:

1. https://github.com/jetpack-io/devbox/blob/main/internal/nix/nixprofile/profile.go#L144
This call is okay, because we have a separate if-condition that handles BinaryCache (i.e. store-paths) installs and is checked before this call.

2. https://github.com/jetpack-io/devbox/blob/main/internal/impl/packages.go#L481
This one in `extraPackagesInProfile` does feel problematic. Because the `item.UnlockedReference` here will be `#` for items-from-store-paths.

Hmm not sure how to fix (2).

Reply:
I'm re-thinking installing via store-path in the profile. Did we go down a wrong path by installing by store-paths in the nix profile?

Downside:
1. we cannot recover a `devpkg.Package` from these item-from-store-paths

Since we switched from content-addressed paths back to input-addressed-paths, I'm wondering if we should continue with the old way of installing packages in profiles. What do we gain (to remind me) from installing by store-path? In theory, it should be the same I believe.

WDYT?

Reply:
Okay, looking at the PR where I made install happen via nix store path: https://github.com/jetpack-io/devbox/pull/1255

the reason we must do it is to avoid downloading nixpkgs. Cool, we need it!

Reply:
Okay, to fix (2), I think we can do something like:
```
	extras := []*nixprofile.NixProfileListItem{}
	// Note: because nix.Input uses memoization when normalizing attribute paths (slow operation),
	// and since we're reusing the Input objects, this O(n*m) loop becomes O(n+m) wrt the slow operation.
outer:
	for _, item := range profileItems {
		profileInput := item.ToPackage(d.lockfile)
		for _, devboxInput := range devboxInputs {
		       // assuming devpkg.FillNarInfoCache() is already called
		       if devboxInput.IsInBinaryCache() {
		          for _, path := range item.storePaths {
		            if devboxInput.InputAddressedPath() == path {
		                // mark as present
		            }
		          }
		       }
		
			if profileInput.Equals(devboxInput) {
				continue outer
			}
		}
		extras = append(extras, item)
	}
```



WDYT?

Reply:
While testing the upgrade path, I concluded that this is an important bug to fix (in separate PR). That's because `tidyProfile` will run the pairwise checks to match profile items with devbox packages every time the number of items in the profile differs from the number of items in devbox.json. And thus, every single time we run into this condition (which I think will be often), we'll end up _removing all store path-installed items from the profile_, because none of them will ever match with what's in devbox.json